### PR TITLE
MSTEST0015: Test method should not be ignored

### PR DESCRIPTION
**Pull request by engyebrahim**
fix: https://github.com/microsoft/testfx/issues/2235

### CODE REVIEW

File: src/Analyzers/MSTest.Analyzers/Resources.resx [hunk 1]
```diff
@@ -448,6 +448,15 @@
   <data name="TestMethodShouldBeValidTitle" xml:space="preserve">
     <value>Test methods should have valid layout</value>
   </data>
+  <data name="TestMethodShouldNotBeIgnoredAnalyzerDescription" xml:space="preserve">
+    <value>Test methods should not be ignored (marked with [Ignore]).</value>
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/Resources.resx]: Nit: let's try to have types or names between single quotes

```suggestion
    <value>Test methods should not be ignored (marked with '[Ignore]').</value>
```

File: src/Analyzers/MSTest.Analyzers/Resources.resx [hunk 2]
```diff
@@ -448,6 +448,15 @@
...
+  </data>
+  <data name="TestMethodShouldNotBeIgnoredAnalyzerFormat" xml:space="preserve">
+    <value>Test Method {0} should not be ignored</value>
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/Resources.resx]: Same nit:

```suggestion
    <value>Test method '{0}' should not be ignored</value>
```

File: src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        DiagnosticIds.TestMethodShouldNotBeIgnoredRuleId,
+        Title,
+        MessageFormat,
+        Description,
+        Category.Design,
+        DiagnosticSeverity.Info,
+        isEnabledByDefault: true);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+    = ImmutableArray.Create(TestMethodShouldNotBeIgnoredRule);
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs]: Nit

```suggestion
    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
        = ImmutableArray.Create(TestMethodShouldNotBeIgnoredRule);
```

File: src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                    SymbolKind.Method);
+            }
+        });
+    }
+
+    private static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol ignoreAttributeSymbol, INamedTypeSymbol testMethodAttributeSymbol)
+    {
+        var methodSymbol = (IMethodSymbol)context.Symbol;
+        if (!methodSymbol.GetAttributes().Any(attr => SymbolEqualityComparer.Default.Equals(attr.AttributeClass, testMethodAttributeSymbol))
+            || !methodSymbol.GetAttributes().Any(attr => SymbolEqualityComparer.Default.Equals(attr.AttributeClass, ignoreAttributeSymbol)))
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs]:
```suggestion
        var methodAttributes = methodSymbol.GetAttributes();
        if (!methodAttributes.Any(attr => SymbolEqualityComparer.Default.Equals(attr.AttributeClass, testMethodAttributeSymbol))
            || !methodAttributesAny(attr => SymbolEqualityComparer.Default.Equals(attr.AttributeClass, ignoreAttributeSymbol)))
```

File: test/UnitTests/MSTest.Analyzers.UnitTests/TestMethodShouldNotBeIgnoredAnalyzerTests.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    public async Task UsingIgnoreWithoutTestMethod_NoDiagnostic()
+    {
+        var code = """
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+            [TestClass]
+            public class MyTestClass
+            {
+                [Ignore]
```
Comment (Evangelink) [inline test/UnitTests/MSTest.Analyzers.UnitTests/TestMethodShouldNotBeIgnoredAnalyzerTests.cs:42]: I don't recall, have we implemented a rule that raises in this case? If not probably worth adding the point to the big ticket with list of rules.

**Review by Evangelink — APPROVED** [review]
Nice work @engyebrahim! A few nits but LGTM.

File: src/Analyzers/MSTest.Analyzers/AnalyzerReleases.Unshipped.md [hunk 6]
```diff
@@ -12,3 +12,4 @@ MSTEST0011 | Usage | Warning | ClassCleanupShouldBeValidAnalyzer, [Documentation
 MSTEST0012 | Usage | Warning | AssemblyInitializeShouldBeValidAnalyzer, [Documentation](https://learn.microsoft.com/dotnet/core/testing/mstest-analyzers/mstest0012)
 MSTEST0013 | Usage | Warning | AssemblyCleanupShouldBeValidAnalyzer, [Documentation](https://learn.microsoft.com/dotnet/core/testing/mstest-analyzers/mstest0013)
 MSTEST0014 | Usage | Warning | DataRowShouldBeValidAnalyzer, [Documentation](https://learn.microsoft.com/dotnet/core/testing/unit-testing-mstest-analyzers-MSTEST0014)
+MSTEST0015 | Design | Info | TestMethodShouldNotBeIgnoredAnalyzer
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/AnalyzerReleases.Unshipped.md]:
```suggestion
MSTEST0015 | Design | Info | TestMethodShouldNotBeIgnoredAnalyzer, [Documentation](https://learn.microsoft.com/dotnet/core/testing/unit-testing-mstest-analyzers-MSTEST0015)
```

File: test/UnitTests/MSTest.Analyzers.UnitTests/TestMethodShouldNotBeIgnoredAnalyzerTests.cs (hunk 5 above)
Comment (engyebrahim) [reply test/UnitTests/MSTest.Analyzers.UnitTests/TestMethodShouldNotBeIgnoredAnalyzerTests.cs:42]: when exactly we have to raise? test class without a test method?

Comment (nohwnd) [reply test/UnitTests/MSTest.Analyzers.UnitTests/TestMethodShouldNotBeIgnoredAnalyzerTests.cs:42]: I think amaury means this rule, that checks that additional test attributes  (Like TestCategory),are only used on methods that have [TestMethod]. And I don't see Ignore in that list of attributes that are checked.

https://learn.microsoft.com/en-us/dotnet/core/testing/mstest-analyzers/mstest0007

File: src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                    SymbolKind.Method);
+            }
+        });
+    }
+
+    private static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol ignoreAttributeSymbol, INamedTypeSymbol testMethodAttributeSymbol)
+    {
+        var methodSymbol = (IMethodSymbol)context.Symbol;
+        var methodAttributes = methodSymbol.GetAttributes();
+        if (!methodAttributes.Any(attr => SymbolEqualityComparer.Default.Equals(attr.AttributeClass, testMethodAttributeSymbol))
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs]: Oh sorry I missed that yesterday, this should be looking for inheritance, as people can inherit from `TestMethod`.

File: test/UnitTests/MSTest.Analyzers.UnitTests/TestMethodShouldNotBeIgnoredAnalyzerTests.cs (hunk 5 above)
Comment (Evangelink) [reply test/UnitTests/MSTest.Analyzers.UnitTests/TestMethodShouldNotBeIgnoredAnalyzerTests.cs:42]: Yes I was referring to that rule. I have been writing too many rules proposal lately and couldn't remember if it was already implemented. `[Ignore]` can also be set on test classes so I am not sure if that's better to update the rule or create a separate rule. Anyways that's out of this PR.

File: src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            }
+        });
+    }
+
+    private static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol ignoreAttributeSymbol, INamedTypeSymbol testMethodAttributeSymbol)
+    {
+        var methodSymbol = (IMethodSymbol)context.Symbol;
+        var methodAttributes = methodSymbol.GetAttributes();
+        bool isTestMethod = false;
+        List<AttributeData> dataRowAttributes = new();
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs]: Unused variable.

File: src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        foreach (var methodAttribute in methodAttributes)
+        {
+            // Current method should be a test method or should inherit from the TestMethod attribute.
+            if (methodAttribute.AttributeClass.Inherits(testMethodAttributeSymbol))
+            {
+                isTestMethod = true;
+            }
+        }
+
+        if (!isTestMethod || !methodAttributes.Any(attr => SymbolEqualityComparer.Default.Equals(attr.AttributeClass, ignoreAttributeSymbol)))
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs]: You are looping above so it's better to also check the ignore attribute above so we don't have to loop twice the same colelction.

File: src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    private static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol ignoreAttributeSymbol, INamedTypeSymbol testMethodAttributeSymbol)
+    {
+        var methodSymbol = (IMethodSymbol)context.Symbol;
+        var methodAttributes = methodSymbol.GetAttributes();
+        bool isTestMethod = false;
+        List<AttributeData> dataRowAttributes = new();
+        foreach (var methodAttribute in methodAttributes)
+        {
+            // Current method should be a test method or should inherit from the TestMethod attribute.
+            if (methodAttribute.AttributeClass.Inherits(testMethodAttributeSymbol))
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/TestMethodShouldNotBeIgnoredAnalyzer.cs:59]: Would be good to have some new tests that checks this is working with inherited.

**Review by Evangelink — APPROVED** [review]

**Pull request merged**